### ISSUE TITLE
Output Shapes of Model During Training and Evaluation Differs

### ISSUE THREAD

**Issue by bmmtstb** [KaiyangZhou/deep-person-reid#558]
I'm using a pretrained osnet_ain model to predict Re-ID embeddings of a specific size from a batch of color images.
But I stumbled upon a problem with the output shapes while trying to predict new values during evaluation time. Consider following example:

```python
import torch
import torch.nn as nn
from torchreid.reid.models.osnet_ain import osnet_ain_x1_0
with torch.no_grad():
    model = osnet_ain_x1_0(num_classes=32, pretrained=True)
    # model = osnet_ain_x1_0(num_classes=32, pretrained=True).eval()
    r = model(torch.randn((1, 3, 256, 128)))  # random "image"
    print(r.shape)
```
For the sake of simplicity, I want a 32-dimensional Re-ID vector for every detection in this batch. I can do so by passing `num_classes=32` to the model. The example above runs fine as long as I simply use the model to predict and dont use the commented eval() line instead.
I am under the assumption that I have to use torch's `no_grad()` and `eval()` to ensure correct behavior for evaluation or if I don't want to train new params.
But during evaluation time (with `eval()`) the model just returns the features of shape [B x 512] and skips the last linear classification layer completely. This means that the forward output now has a fixed size of 512 and is independant of the `num_classes=32` I set initially.

The Linear layer is set up correctly during init:
https://github.com/KaiyangZhou/deep-person-reid/blob/566a56a2cb255f59ba75aa817032621784df546a/torchreid/models/osnet_ain.py#L363

But in the forward pass, the model skips the classifier when it is not training:
https://github.com/KaiyangZhou/deep-person-reid/blob/566a56a2cb255f59ba75aa817032621784df546a/torchreid/models/osnet_ain.py#L432-L446

Am I missing something? For me different output shapes during training and evaluation is not expected behavior.
I could simply call model.classifier(r) after, but this doesn't feel right...

**Comment by bmmtstb** [comment]
It took a while, but I think I understood now. I write a few notes, if someone else stumbles upon the same issues, and close this issue.

Looking at the engine, e.g. the ``ImageSoftmaxEngine`` during the training process in forward_backward() the loss is computed between the pids and the models prediction.

https://github.com/KaiyangZhou/deep-person-reid/blob/566a56a2cb255f59ba75aa817032621784df546a/torchreid/engine/image/softmax.py#L86

In this case the model prediction is a tensor of shape ``[(B,) num_classes]`` containing an estimate of a certainty for every class. (This is not really a probability, because it wasn't passed through a Softmax, but the analogy holds.) Still, by passing it to the loss function, the last linear layer will learn to predict the class from the embeddings. And because during training there are no "real" embeddings to compute a loss against, it is important to predict the classes to be able to compare those to the targets.

---

During testing, we assume that our model predicts useful embeddings and therefore we can use them. Lets have a look at engine._evaluate, when calling ``_feature_extraction()``:

https://github.com/KaiyangZhou/deep-person-reid/blob/566a56a2cb255f59ba75aa817032621784df546a/torchreid/engine/engine.py#L379-L385

The same embedding generation model will predict the embeddings (`qf` & `gf`) for the query and gallery. Then we can validate, whether the embeddings for person X in the query are close to the embeddings for person X in the gallery. This is done by computing the distance between every embedding in the query to every embedding in the gallery. Smaller distances meaning that the embeddings are closer, and therefore it is more likely, that the same person is shown. If the embeddings for each individual are close, and still far away from every other individual, the goal is reached.

Looking at the evaluation (here the python version of eval_market1501), we first obtain the indices which sort the distance matrix (smallest to highest).

https://github.com/KaiyangZhou/deep-person-reid/blob/566a56a2cb255f59ba75aa817032621784df546a/torchreid/metrics/rank.py#L107-L108

Therefore, with ``g_pids[indices]`` we obtain the for every person id in the gallery a sorted list of the closest matching person ids from query. If we now take the first `max_rank` values of the matches, which contain the most likely classes, we can check if those contain the correct target id.

---

So, Yes, it kind of makes sense to have different outputs. But in my opinion it would be better to always return the embeddings and a probability distribution (e.g. Softmax) of the certainty that this embedding relates to a specific class. Then there wouldn't have to be a difference between training and testing...
